### ISSUE TITLE
Increase test coverage

### ISSUE THREAD

**Issue by czechboy0** [czechboy0/XcodeServerSDK#45]
Right now our code coverage is somewhere around 15%. :zap: 

![](http://i963.photobucket.com/albums/ae115/shineeluv31/gif/theofficeno.gif)

**Comment by cojoj** [comment]
Well, that's not good but I there are some methods in which I see no sense of covering them with unit tests. What's more, some methods are't totally clear for me because of their dependencie, so I skip them as I cn't write propert tests for them 😭

---

I was thinking this morning about integrating [Travis-CI](https://travis-ci.org). What'd you say @czechboy0?

---

Also, found this cool thing called [coveralls](https://coveralls.io) 😃

**Comment by czechboy0** [comment]
They [still don't have](http://next.travis-ci.com) Xcode 7 support.

**Comment by cojoj** [comment]
![](http://media3.giphy.com/media/g9Bgr1cPc3L68/giphy.gif)

**Comment by czechboy0** [comment]
Their sluggish new Xcode version support is the whole reason Buildasaur exists :laughing:

**Comment by czechboy0** [comment]
But yeah once Travis has Xcode 7 support, I'd be up for using it for testing and coverage. No idea how long that will take however.

**Comment by czechboy0** [comment]
In the meantime we have Xcode 7's code coverage metric, so run it locally and report it in each PR, so that we know what we're at.

**Comment by cojoj** [comment]
There's no school like the oldschool... 😆

**Comment by cojoj** [comment]
I've seen test coverage for `XcodeServer` but all of them have `DEV_` prefix so they're not rally doing anything at the moment. @czechboy0 is this on purpose? I wonder if DVR integration is ready and I can use it to implement new test cases?

**Comment by czechboy0** [comment]
Yes, the DEV_ prefix is there on purpose. Those are live tests that actually affect and call the real server and I use it during development. So they need to be disabled most of the time unless you're testing live API calls. 

And yes, DVR is ready. Just run carthage update --no-build and you're all set. I also added a folder just for Cassettes in the test target/folder.

**Comment by cojoj** [comment]
Hmm, any possibility you write steps needed to use DVR? I how now idea where to put my hands on 😞

**Comment by czechboy0** [comment]
Look at test DEV_testLive_FetchAndRecordBot() in XcodeServerTests. 

Create a test like that for whatever you want to test. Then you run it once and it records and crashes, printing out where it has saved the cassette. You copy that file into the test target in the Cassettes folder, add it to the project **to both test targets only** and then run it again. This time it will run fine, using the loaded cassette.

**Comment by cojoj** [comment]
Great, but it's easy for `GET`s, so how about `POST`s? Do I have to manually remove added entities?

**Comment by cojoj** [comment]
With #48 we get **25%** of test coverage for `XcodeServerSDK`... Slowly getting more and more 👍

**Comment by esttorhe** [comment]
Have we checked asking for [`CircleCI`](https://circleci.com) access instead of `Travis` maybe?

**Comment by cojoj** [comment]
Isn't **CirlceCI** paid?

**Comment by czechboy0** [comment]
Yeah I tried them all yesterday. They don't yet support Xcode 7 :( Classic.

**Comment by cojoj** [comment]
**Travis** today has confirmed official support from 6.4...

**Comment by czechboy0** [comment]
Link? But that doesn't help anyway. We need Xcode 7. Hopefully I'll be able to release an experimental build of Buildasaur soon locally on my Mac to get PR testing again.

**Comment by cojoj** [comment]
<blockquote class="twitter-tweet" lang="pl"><p lang="en" dir="ltr">We just added support for Xcode 6.4 and the iOS 8.4 SDK! Add `osx_image: xcode6.4` to your .travis.yml to begin using it.</p>&mdash; Travis CI (@travisci) <a href="https://twitter.com/travisci/status/616007115832999936">czerwiec 30, 2015</a></blockquote> <script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

**Comment by esttorhe** [comment]
_CircleCI_ is free if we use it only for 1 job at a time (which i think would be fine) except there's no `Xcode 7` love yet.

---

how about [`Bitrise`](https://bitrise.io) ? (just checked, they support up to 6.4)

**Comment by cojoj** [comment]
I guess we'll have to wait for anything to support `Xcode 7`... Other possibility - someone has to allow us to access `OS X Server 5`.

**Comment by esttorhe** [comment]
We can use [`Ship.io`](https://ship.io); they have `Xcode 7` support already https://ship.io/xcode-7-beta-is-here/

They have a free plan for individual accounts with 2 jobs or something along those lines; might be worth looking at it

**Comment by cojoj** [comment]
👌

**Comment by czechboy0** [comment]
Just FYI, I do have an Xcode Bot running locally which tests the `swift-2` branch after every single commit, so no worries there. I turned on email notifications so you'll also get them if something breaks.

Unfortunately, we need Pull Request testing, which [Buildasaur](https://github.com/czechboy0/Buildasaur) has (I'm working on Xcode 7 support there) and CI services like Travis do as well, but they don't yet support Xcode 7. It's exactly what happened last year. Ship.io doesn't give us any benefit over a local Xcode Bot. Plus, their Scheme detection is broken, I tried it yesterday.

So once Buildasaur has an Xcode 7 version ready I'll turn it on for this repo so that all PRs will get tested.

**Comment by esttorhe** [comment]
:clap: that sounds great then :tada: 

TBH i didn't like Ship.io when I tried it.

If I can help with bringing Xcode 7 support to Buildasaur just let me know

**Comment by esttorhe** [comment]
Am I the only one seeing weird behavior with code coverage?

Yesterday I added a test that specifically triggers the invalid scheme path of code but this is what i'm seeing after running that tests:
![screen shot 2015-07-01 at 15 31 27](https://cloud.githubusercontent.com/assets/2642850/8465310/4d473c2c-2006-11e5-8d31-f483a66260b3.png)

Is marked as never called

**Comment by cojoj** [comment]
That's really strange... Was it covering the failable path yesterday? Looks fine to me...

Nothing has changed in those files AFAIK.

**Comment by esttorhe** [comment]
It wasn't covered and that's why I went and wrote a test for it but didn't check. now I came to check and is still uncovered.

No matter if I run only that test or all the tests on iOS or OSX that path keeps being marked as not checked.

Perhaps the `throws` thing has something to do with it?

**Comment by esttorhe** [comment]
Disregard my last comment… this invalidates my theory

``` swift
guard let url = NSURL(string: host) else {
            /*******************************************************************
             **   Had to be added to silence the compiler ¯\_(ツ)_/¯
             **   Radar: http://openradar.me/21514477
             **   Reply: https://twitter.com/jckarter/status/613491369311535104
             ******************************************************************/
            self.host = ""; self.user = nil; self.password = nil

            throw ConfigurationErrors.InvalidHostProvided(host)
        }
```

This is passing and marked as checked

**Comment by czechboy0** [comment]
Just FYI, current state of things: 
![screenshot 2015-07-01 23 37 05](https://cloud.githubusercontent.com/assets/2182121/8466392/279671b6-204a-11e5-8847-a725a932b810.png)

**Comment by esttorhe** [comment]
That's weird; if you check at the tests and the coverage `XcodeServerConfig` should be near 100% but the coverage is not working as expected

**Comment by czechboy0** [comment]
It might also have bugs. We'll see throughout the summer. We know we're still not testing most of the parsing/dictionarifying logic anyway :) I guess once we're confident about that we can tackle these weirdnesses.

**Comment by cojoj** [comment]
Hmmm this whole **test coverage** works fine for me everytime but bugs are possible...

**Comment by cojoj** [comment]
@czechboy0, @esttorhe I wonder...
**Who's the first one to strat writing tests for API calls?** 😜

**Comment by czechboy0** [comment]
All the tools are there. [Here](https://github.com/czechboy0/XcodeServerSDK/blob/5be5118efc3eea7d73b249ba34d8a27be72eb744/XcodeServerSDKTests/XcodeServerTests.swift#L34) is an example of a working test and how easily you get a recording Server. And [here](https://github.com/czechboy0/XcodeServerSDK/blob/5be5118efc3eea7d73b249ba34d8a27be72eb744/XcodeServerSDKTests/TestUtils.swift#L27) is how it works.

**Comment by czechboy0** [comment]
And [here](https://github.com/czechboy0/XcodeServerSDK/blob/5be5118efc3eea7d73b249ba34d8a27be72eb744/XcodeServerSDKTests/BotParsingTests.swift#L32) is how you can reuse a recorded `Cassette` to add many parsing tests.

**Comment by cojoj** [comment]
Did you just point at me? 😜

**Comment by czechboy0** [comment]
:point_left: :point_right: :point_left: :point_right: :point_down: :point_up: :point_down: :point_up: 

Not pointing at anyone, just :dancer:

**Comment by cojoj** [comment]
![](http://media0.giphy.com/media/ObZEj9rGiC7Ru/giphy.gif)

**Comment by cojoj** [comment]
On thing I've noticed, you use:

``` swift
let config = try! XcodeServerConfig(
            host: "https://127.0.0.1",
            user: "ICanCreateBots",
            password: "superSecr3t")
```

For sake of consistency, should we use default `XcodeServerConfig` or pass ours? I think the first option will do better as all tests will have the same `XcodeServerConfig` but in this case all of us will have to create _ICanCreateBots_ user with _superSecr3t_ password.

**Comment by czechboy0** [comment]
Depends how your XCS is setup, but I have that anyone can view bots and only logged in users can create bots. Which means that any user, including the non-existent ICanCreateBots will return all read-only requests. So you don't need to create this user at all to test most get requests. 

When it comes to post requests, I haven't decided yet how we're going to test that since in post requests it's what we're _sending_ is what we want to test, whereas with get requests it's what you're _receiving_ that you want to test. 

TL;DR - enable any user to read bots on your XCS and let's write tests for all the get requests first. I'll think of a good way to test posts later. Aight?

**Comment by cojoj** [comment]
I can't agree with you... While testing `POST` requests it's necessary to assert **request** but **response** is also important!

**Comment by cojoj** [comment]
@czechboy0 strange thing happened... I've written test for `getRepository()` which looks like:

``` swift
func testGetRepositories() {
        let expectation = self.expectationWithDescription("Get Repositories")
        let server = self.getRecordingXcodeServer("get_repositories")

        server.getRepositories() { (repositories, error) in
            XCTAssertNil(error, "Error should be nil")
            XCTAssertNotNil(repositories, "Repositories shouldn't be nil")

            if let repos = repositories {
                XCTAssertEqual(repos.count, 2, "There should be two repositories available")

                for (index, repo) in repos.enumerate() {
                    XCTAssertEqual(repo.name, "Test\(index + 1)")
                }
            }

            expectation.fulfill()
        }

        self.waitForExpectationsWithTimeout(10) { error in
            if let error = error {
                print("Timeout error: \(error.localizedDescription)")
            }
        }
    }
```

So at first I had to move `get_repositories.json` and re-run tests - fine, works and even test is passing but when I look at code coverage I'm like **WTF, bro!?**:
![screen shot 2015-07-02 at 14 53 00](https://cloud.githubusercontent.com/assets/4522407/8477399/1ca85dca-20ca-11e5-9a60-2c67b21ce09d.png)

I can agree that the fragment where `completion` with `nil` is passed may not be executed but `[Repository]` parsing?

---

I set a breakpoint where the code _isn't covered_ and it executes so this is being called but marked differently 😓

**Comment by cojoj** [comment]
@esttorhe seems like I've been having similiar kind of issue with `guard` statements...

**Comment by esttorhe** [comment]
Yeah; I saw your tweet to Joe; seems like radar time!! :dancers:

**Comment by cojoj** [comment]
![](http://images.lazygamer.net/2013/07/sad-but-true.jpg)

**Comment by cojoj** [comment]
New follower...

![joe](https://cloud.githubusercontent.com/assets/4522407/8487249/1ab69e50-210b-11e5-8ee4-860d01101a48.png)

**Things just got serious**! 😁

**Comment by czechboy0** [comment]
:fireworks:

**Comment by czechboy0** [comment]
Getting around 50% now. Not great but getting much better, especially thanks to @cojoj!

**Comment by czechboy0** [comment]
This is nice, we had fun with it, but it's not really actionable. Closing.
